### ISSUE TITLE
Balancing Folds - Donnacha Oisín Kidney

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7235]
Balancing Folds - Donnacha Ois&iacute;n Kidney<br>
<br>
https://ift.tt/2T9CjHt<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<h2>Balancing Folds</h2>
<p>Posted on October 30, 2017</p>
<p>There are three main ways to fold things in Haskell: from the right, from the left, and from either side. Let&rsquo;s look at the left vs right variants first. <code><span>foldr</span></code> works from the right:</p>
<div>
<pre><code><span><a href="#cb1-1"></a><span>foldr</span> (<span>+</span>) <span>0</span> [<span>1</span>,<span>2</span>,<span>3</span>]</span>
<span><a href="#cb1-2"></a><span>1</span> <span>+</span> (<span>2</span> <span>+</span> (<span>3</span> <span>+</span> <span>0</span>))</span></code></pre>
</div>
<p>And <code><span>foldl</span></code> from the left:</p>
<div>
<pre><code><span><a href="#cb2-1"></a><span>foldl</span> (<span>+</span>) <span>0</span> [<span>1</span>,<span>2</span>,<span>3</span>]</span>
<span><a href="#cb2-2"></a>((<span>0</span> <span>+</span> <span>1</span>) <span>+</span> <span>2</span>) <span>+</span> <span>3</span></span></code></pre>
</div>
<p>As you&rsquo;ll notice, the result of the two operations above is the same (6; although one may take much longer than the other). In fact, <em>whenever</em> the result of <code><span>foldr</span></code> and <code><span>foldl</span></code> is the same for a pair of arguments (in this case <code><span>+</span></code> and <code><span>0</span></code>), we say that that pair forms a <a href="https://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Monoid.html#t:Monoid"><code><span>Monoid</span></code></a> for some type (well, there&rsquo;s some extra stuff to do with <code><span>0</span></code>, but I only care about associativity at the moment). In this case, the <a href="https://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Monoid.html#t:Sum"><code><span>Sum</span></code></a> monoid is formed:</p>
<div>
<pre><code><span><a href="#cb3-1"></a><span>newtype</span> <span>Sum</span> a <span>=</span> <span>Sum</span> {<span> getSum ::</span> a }</span>
<span><a href="#cb3-2"></a></span>
<span><a href="#cb3-3"></a><span>instance</span> <span>Num</span> a <span>=&gt;</span> <span>Monoid</span> (<span>Sum</span> a) <span>where</span></span>
<span><a href="#cb3-4"></a> <span>mempty</span> <span>=</span> <span>Sum</span> <span>0</span></span>
<span><a href="#cb3-5"></a> <span>mappend</span> (<span>Sum</span> x) (<span>Sum</span> y) <span>=</span> <span>Sum</span> (x <span>+</span> y)</span></code></pre>
</div>
<p>When you know that you have a monoid, you can use the <a href="https://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Foldable.html#v:foldMap"><code><span>foldMap</span></code></a> function: this is the third kind of fold. It says that you don&rsquo;t care which of <code><span>foldl</span></code> or <code><span>foldr</span></code> is used, so the implementer of <code><span>foldMap</span></code> can put the parentheses wherever they want:</p>
<div>
<pre><code><span><a href="#cb4-1"></a><span>foldMap</span> <span>Sum</span> [<span>1</span>,<span>2</span>,<span>3</span>]</span>
<span><a href="#cb4-2"></a>(<span>1</span> <span>+</span> <span>2</span>) <span>+</span> (<span>3</span> <span>+</span> <span>0</span>)</span>
<span><a href="#cb4-3"></a><span>0</span> <span>+</span> ((<span>1</span> <span>+</span> <span>2</span>) <span>+</span> <span>3</span>)</span>
<span><a href="#cb4-4"></a>((<span>0</span> <span>+</span> <span>1</span>) <span>+</span> <span>2</span>) <span>+</span> <span>3</span></span></code></pre>
</div>
<p>And we can&rsquo;t tell the difference from the result. This is a pretty bare-bones introduction to folds and monoids: you won&rsquo;t need to know more than that for the rest of this post, but the topic area is fascinating and deep, so don&rsquo;t let me give you the impression that I&rsquo;ve done anything more than scratched the surface.</p>
<p>Quite often, we <em>do</em> care about where the parentheses go. Take, for instance, a binary tree type, with values at the leaves:</p>
<div>
<pre><code><span><a href="#cb5-1"></a><span>data</span> <span>Tree</span> a</span>
<span><a href="#cb5-2"></a> <span>=</span> <span>Empty</span></span>
<span><a href="#cb5-3"></a> <span>|</span> <span>Leaf</span> a</span>
<span><a href="#cb5-4"></a> <span>|</span> <span>Tree</span> a <span>:*:</span> <span>Tree</span> a</span>
<span><a href="#cb5-5"></a></span>
<span><a href="#cb5-6"></a><span>instance</span> <span>Show</span> a <span>=&gt;</span></span>
<span><a href="#cb5-7"></a> <span>Show</span> (<span>Tree</span> a) <span>where</span></span>
<span><a href="#cb5-8"></a> <span>show</span> <span>Empty</span> <span>=</span> <span>"()"</span></span>
<span><a href="#cb5-9"></a> <span>show</span> (<span>Leaf</span> x) <span>=</span> <span>show</span> x</span>
<span><a href="#cb5-10"></a> <span>show</span> (l <span>:*:</span> r) <span>=</span> <span>"("</span> <span>++</span> <span>show</span> l <span>++</span> <span>"*"</span> <span>++</span> <span>show</span> r <span>++</span> <span>")"</span></span></code></pre>
</div>
<p>We can&rsquo;t (well, shouldn&rsquo;t) us <code><span>foldMap</span></code> here, because we would be able to tell the difference between different arrangements of parentheses:</p>
<div>
<pre><code><span><a href="#cb6-1"></a><span>&gt;&gt;&gt;</span> <span>foldMap</span> something [<span>1</span>,<span>2</span>,<span>3</span>]</span>
<span><a href="#cb6-2"></a></span>
<span><a href="#cb6-3"></a>((<span>1</span><span>*</span><span>2</span>)<span>*</span>(<span>3</span><span>*</span>())) &#9474; (()<span>*</span>((<span>1</span><span>*</span><span>2</span>)<span>*</span><span>3</span>)) &#9474; (((()<span>*</span><span>1</span>)<span>*</span><span>2</span>)<span>*</span><span>3</span>)</span>
<span><a href="#cb6-4"></a>&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9532;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9532;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;&#9472;</span>
<span><a href="#cb6-5"></a> &#9484;<span>1</span> &#9474; &#9484;() &#9474; &#9484;()</span>
<span><a href="#cb6-6"></a> &#9484;&#9508; &#9474; &#9508; &#9474; &#9484;&#9508;</span>
<span><a href="#cb6-7"></a> &#9474;&#9492;<span>2</span> &#9474; &#9474; &#9484;<span>1</span> &#9474; &#9474;&#9492;<span>1</span></span>
<span><a href="#cb6-8"></a> &#9508; &#9474; &#9474;&#9484;&#9508; &#9474; &#9484;&#9508;</span>
<span><a href="#cb6-9"></a> &#9474;&#9484;<span>3</span> &#9474; &#9474;&#9474;&#9492;<span>2</span> &#9474; &#9474;&#9492;<span>2</span></span>
<span><a href="#cb6-10"></a> &#9492;&#9508; &#9474; &#9492;&#9508; &#9474; &#9508;</span>
<span><a href="#cb6-11"></a> &#9492;() &#9474; &#9492;<span>3</span> &#9474; &#9492;<span>3</span></span></code></pre>
</div>
<p>So we use one of the folds which lets us choose the arrangements of parentheses:</p>
<div>
<pre><code><span><a href="#cb7-1"></a><span>&gt;&gt;&gt;</span> (<span>foldr</span> (<span>:*:</span>) <span>Empty</span> <span>.</span> <span>map</span> <span>Leaf</span>) [<span>1</span>,<span>2</span>,<span>3</span>,<span>4</span>,<span>5</span>,<span>6</span>]</span>
<span><a href="#cb7-2"></a>(<span>1</span><span>*</span>(<span>2</span><span>*</span>(<span>3</span><span>*</span>(<span>4</span><span>*</span>(<span>5</span><span>*</span>(<span>6</span><span>*</span>()))))))</span>
<span><a href="#cb7-3"></a> &#9484;<span>1</span></span>
<span><a href="#cb7-4"></a> &#9484;&#9508;</span>
<span><a href="#cb7-5"></a> &#9474;&#9492;<span>2</span></span>
<span><a href="#cb7-6"></a> &#9484;&#9508;</span>
<span><a href="#cb7-7"></a> &#9474;&#9492;<span>3</span></span>
<span><a href="#cb7-8"></a> &#9484;&#9508;</span>
<span><a href="#cb7-9"></a> &#9474;&#9492;<span>4</span></span>
<span><a href="#cb7-10"></a> &#9484;&#9508;</span>
<span><a href="#cb7-11"></a> &#9474;&#9492;<span>5</span></span>
<span><a href="#cb7-12"></a>&#9484;&#9508;</span>
<span><a href="#cb7-13"></a>&#9474;&#9492;<span>6</span></span>
<span><a href="#cb7-14"></a>&#9508;</span>
<span><a href="#cb7-15"></a>&#9492;()</span>
<span><a href="#cb7-16"></a></span>
<span><a href="#cb7-17"></a><span>&gt;&gt;&gt;</span> (<span>foldl</span> (<span>:*:</span>) <span>Empty</span> <span>.</span> <span>map</span> <span>Leaf</span>) [<span>1</span>,<span>2</span>,<span>3</span>,<span>4</span>,<span>5</span>,<span>6</span>]</span>
<span><a href="#cb7-18"></a>((((((()<span>*</span><span>1</span>)<span>*</span><span>2</span>)<span>*</span><span>3</span>)<span>*</span><span>4</span>)<span>*</span><span>5</span>)<span>*</span><span>6</span>)</span>
<span><a href="#cb7-19"></a>&#9484;()</span>
<span><a href="#cb7-20"></a>&#9508;</span>
<span><a href="#cb7-21"></a>&#9474;&#9484;<span>1</span></span>
<span><a href="#cb7-22"></a>&#9492;&#9508;</span>
<span><a href="#cb7-23"></a> &#9474;&#9484;<span>2</span></span>
<span><a href="#cb7-24"></a> &#9492;&#9508;</span>
<span><a href="#cb7-25"></a> &#9474;&#9484;<span>3</span></span>
<span><a href="#cb7-26"></a> &#9492;&#9508;</span>
<span><a href="#cb7-27"></a> &#9474;&#9484;<span>4</span></span>
<span><a href="#cb7-28"></a> &#9492;&#9508;</span>
<span><a href="#cb7-29"></a> &#9474;&#9484;<span>5</span></span>
<span><a href="#cb7-30"></a> &#9492;&#9508;</span>
<span><a href="#cb7-31"></a> &#9492;<span>6</span></span></code></pre>
</div>
<p>The issue is that neither of the trees generated are necessarily what we want: often, we want something more <em>balanced</em>.</p>
<h2>TreeFold</h2>
<p>To try and find a more balanced fold, let&rsquo;s (for now) assume we&rsquo;re always going to get non-empty input. This will let us simplify the <code><span>Tree</span></code> type a little, to:</p>
<div>
<pre><code><span><a href="#cb8-1"></a><span>data</span> <span>Tree</span> a</span>
<span><a href="#cb8-2"></a> <span>=</span> <span>Leaf</span> a</span>
<span><a href="#cb8-3"></a> <span>|</span> <span>Tree</span> a <span>:*:</span> <span>Tree</span> a</span>
<span><a href="#cb8-4"></a> <span>deriving</span> <span>Foldable</span></span>
<span><a href="#cb8-5"></a></span>
<span><a href="#cb8-6"></a><span>instance</span> <span>Show</span> a <span>=&gt;</span></span>
<span><a href="#cb8-7"></a> <span>Show</span> (<span>Tree</span> a) <span>where</span></span>
<span><a href="#cb8-8"></a> <span>show</span> (<span>Leaf</span> x) <span>=</span> <span>show</span> x</span>
<span><a href="#cb8-9"></a> <span>show</span> (l <span>:*:</span> r) <span>=</span> <span>"("</span> <span>++</span> <span>show</span> l <span>++</span> <span>"*"</span> <span>++</span> <span>show</span> r <span>++</span> <span>")"</span></span></code></pre>
</div>
<p>Then, we can use Jon Fairbairn&rsquo;s fold described in <a href="http://www.mail-archive.com/haskell@haskell.org/msg01788.html">this</a> email, adapted a bit for our non-empty input:</p>
<div>
<pre><code><span><a href="#cb9-1"></a><span>import</span> <span>Data.List.NonEmpty</span> (<span>NonEmpty</span>(..))</span>
<span><a href="#cb9-2"></a></span>
<span><a href="#cb9-3"></a><span>treeFold ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> <span>NonEmpty</span> a <span>-&gt;</span> a</span>
<span><a href="#cb9-4"></a>treeFold f <span>=</span> go</span>
<span><a href="#cb9-5"></a> <span>where</span></span>
<span><a href="#cb9-6"></a> go (x <span>:|</span> []) <span>=</span> x</span>
<span><a href="#cb9-7"></a> go (a <span>:|</span> b<span>:</span>l) <span>=</span> go (f a b <span>:|</span> pairMap l)</span>
<span><a href="#cb9-8"></a> pairMap (x<span>:</span>y<span>:</span>rest) <span>=</span> f x y <span>:</span> pairMap rest</span>
<span><a href="#cb9-9"></a> pairMap xs <span>=</span> xs</span></code></pre>
</div>
<p>There are two parts to this function: <code>pairMap</code> and the <code>go</code> helper. <code>pairMap</code> combines adjacent elements in the list using the combining function. As a top-level function it might look like this:</p>
<div>
<pre><code><span><a href="#cb10-1"></a>pairMap f (x<span>:</span>y<span>:</span>rest) <span>=</span> f x y <span>:</span> pairMap f rest</span>
<span><a href="#cb10-2"></a>pairMap f xs <span>=</span> xs</span>
<span><a href="#cb10-3"></a></span>
<span><a href="#cb10-4"></a>pairMap (<span>++</span>) [<span>"a"</span>,<span>"b"</span>,<span>"c"</span>,<span>"d"</span>,<span>"e"</span>]</span>
<span><a href="#cb10-5"></a><span>-- ["ab","cd","e"]</span></span></code></pre>
</div>
<p>As you can see, it leaves any leftovers untouched at the end of the list.</p>
<p>The <code>go</code> helper applies <code>pairMap</code> repeatedly to the list until it has only one element. This gives us much more balanced results that <code><span>foldl</span></code> or <code><span>foldr</span></code> (turn on <code><span>-</span><span>XOverloadedLists</span></code> to write non-empty lists using this syntax):</p>
<div>
<pre><code><span><a href="#cb11-1"></a><span>&gt;&gt;&gt;</span> (treeFold (<span>:*:</span>) <span>.</span> <span>fmap</span> <span>Leaf</span>) [<span>1</span>,<span>2</span>,<span>3</span>,<span>4</span>,<span>5</span>,<span>6</span>]</span>
<span><a href="#cb11-2"></a>(((<span>1</span><span>*</span><span>2</span>)<span>*</span>(<span>3</span><span>*</span><span>4</span>))<span>*</span>(<span>5</span><span>*</span><span>6</span>))</span>
<span><a href="#cb11-3"></a> &#9484;<span>1</span></span>
<span><a href="#cb11-4"></a> &#9484;&#9508;</span>
<span><a href="#cb11-5"></a> &#9474;&#9492;<span>2</span></span>
<span><a href="#cb11-6"></a>&#9484;&#9508;</span>
<span><a href="#cb11-7"></a>&#9474;&#9474;&#9484;<span>3</span></span>
<span><a href="#cb11-8"></a>&#9474;&#9492;&#9508;</span>
<span><a href="#cb11-9"></a>&#9474; &#9492;<span>4</span></span>
<span><a href="#cb11-10"></a>&#9508;</span>
<span><a href="#cb11-11"></a>&#9474;&#9484;<span>5</span></span>
<span><a href="#cb11-12"></a>&#9492;&#9508;</span>
<span><a href="#cb11-13"></a> &#9492;<span>6</span></span>
<span><a href="#cb11-14"></a></span>
<span><a href="#cb11-15"></a><span>&gt;&gt;&gt;</span> (treeFold (<span>:*:</span>) <span>.</span> <span>fmap</span> <span>Leaf</span>) [<span>1</span>,<span>2</span>,<span>3</span>,<span>4</span>,<span>5</span>,<span>6</span>,<span>7</span>,<span>8</span>]</span>
<span><a href="#cb11-16"></a>(((<span>1</span><span>*</span><span>2</span>)<span>*</span>(<span>3</span><span>*</span><span>4</span>))<span>*</span>((<span>5</span><span>*</span><span>6</span>)<span>*</span>(<span>7</span><span>*</span><span>8</span>)))</span>
<span><a href="#cb11-17"></a> &#9484;<span>1</span></span>
<span><a href="#cb11-18"></a> &#9484;&#9508;</span>
<span><a href="#cb11-19"></a> &#9474;&#9492;<span>2</span></span>
<span><a href="#cb11-20"></a>&#9484;&#9508;</span>
<span><a href="#cb11-21"></a>&#9474;&#9474;&#9484;<span>3</span></span>
<span><a href="#cb11-22"></a>&#9474;&#9492;&#9508;</span>
<span><a href="#cb11-23"></a>&#9474; &#9492;<span>4</span></span>
<span><a href="#cb11-24"></a>&#9508;</span>
<span><a href="#cb11-25"></a>&#9474; &#9484;<span>5</span></span>
<span><a href="#cb11-26"></a>&#9474;&#9484;&#9508;</span>
<span><a href="#cb11-27"></a>&#9474;&#9474;&#9492;<span>6</span></span>
<span><a href="#cb11-28"></a>&#9492;&#9508;</span>
<span><a href="#cb11-29"></a> &#9474;&#9484;<span>7</span></span>
<span><a href="#cb11-30"></a> &#9492;&#9508;</span>
<span><a href="#cb11-31"></a> &#9492;<span>8</span></span></code></pre>
</div>
<p>However, there are still cases where one branch will be much larger than its sibling. The fold fills a balanced binary tree from the left, but any leftover elements are put at the top level. In other words:</p>
<div>
<pre><code><span><a href="#cb12-1"></a><span>&gt;&gt;&gt;</span> (treeFold (<span>:*:</span>) <span>.</span> <span>fmap</span> <span>Leaf</span>) [<span>1</span><span>..</span><span>9</span>]</span>
<span><a href="#cb12-2"></a>((((<span>1</span><span>*</span><span>2</span>)<span>*</span>(<span>3</span><span>*</span><span>4</span>))<span>*</span>((<span>5</span><span>*</span><span>6</span>)<span>*</span>(<span>7</span><span>*</span><span>8</span>)))<span>*</span><span>9</span>)</span>
<span><a href="#cb12-3"></a> &#9484;<span>1</span></span>
<span><a href="#cb12-4"></a> &#9484;&#9508;</span>
<span><a href="#cb12-5"></a> &#9474;&#9492;<span>2</span></span>
<span><a href="#cb12-6"></a> &#9484;&#9508;</span>
<span><a href="#cb12-7"></a> &#9474;&#9474;&#9484;<span>3</span></span>
<span><a href="#cb12-8"></a> &#9474;&#9492;&#9508;</span>
<span><a href="#cb12-9"></a> &#9474; &#9492;<span>4</span></span>
<span><a href="#cb12-10"></a>&#9484;&#9508;</span>
<span><a href="#cb12-11"></a>&#9474;&#9474; &#9484;<span>5</span></span>
<span><a href="#cb12-12"></a>&#9474;&#9474;&#9484;&#9508;</span>
<span><a href="#cb12-13"></a>&#9474;&#9474;&#9474;&#9492;<span>6</span></span>
<span><a href="#cb12-14"></a>&#9474;&#9492;&#9508;</span>
<span><a href="#cb12-15"></a>&#9474; &#9474;&#9484;<span>7</span></span>
<span><a href="#cb12-16"></a>&#9474; &#9492;&#9508;</span>
<span><a href="#cb12-17"></a>&#9474; &#9492;<span>8</span></span>
<span><a href="#cb12-18"></a>&#9508;</span>
<span><a href="#cb12-19"></a>&#9492;<span>9</span></span></code></pre>
</div>
<p>That <code><span>9</span></code> hanging out on its own there is a problem.</p>
<h2>Typewriters and Slaloms</h2>
<p>One observation we can make is that <code>pairMap</code> always starts from the same side on each iteration, like a typewriter moving from one line to the next. This has the consequence of building up the leftovers on one side, leaving them until the top level.</p>
<p>We can improve the situation slightly by going back and forth, slalom-style, so we consume leftovers on each iteration:</p>
<div>
<pre><code><span><a href="#cb13-1"></a><span>treeFold ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> <span>NonEmpty</span> a <span>-&gt;</span> a</span>
<span><a href="#cb13-2"></a>treeFold f <span>=</span> goTo <span>where</span></span>
<span><a href="#cb13-3"></a> </span>
<span><a href="#cb13-4"></a> goTo (y <span>:|</span> []) <span>=</span> y</span>
<span><a href="#cb13-5"></a> goTo (a <span>:|</span> b <span>:</span> rest) <span>=</span> goFro (pairMap f (f a b) rest)</span>
<span><a href="#cb13-6"></a> goFro (y <span>:|</span> []) <span>=</span> y</span>
<span><a href="#cb13-7"></a> goFro (a <span>:|</span> b <span>:</span> rest) <span>=</span> goTo (pairMap (<span>flip</span> f) (f b a) rest)</span>
<span><a href="#cb13-8"></a></span>
<span><a href="#cb13-9"></a> pairMap f <span>=</span> go [] <span>where</span></span>
<span><a href="#cb13-10"></a> go ys y (a<span>:</span>b<span>:</span>rest) <span>=</span> go (y<span>:</span>ys) (f a b) rest</span>
<span><a href="#cb13-11"></a> go ys y [z] <span>=</span> z <span>:|</span> y <span>:</span> ys</span>
<span><a href="#cb13-12"></a> go ys y [] <span>=</span> y <span>:|</span> ys</span></code></pre>
</div>
<p>Notice that we have to flip the combining function to make sure the ordering is the same on output. For the earlier example, this solves the issue:</p>
<div>
<pre><code><span><a href="#cb14-1"></a><span>&gt;&gt;&gt;</span> (treeFold (<span>:*:</span>) <span>.</span> <span>fmap</span> <span>Leaf</span>) [<span>1</span><span>..</span><span>9</span>]</span>
<span><a href="#cb14-2"></a>(((<span>1</span><span>*</span><span>2</span>)<span>*</span>((<span>3</span><span>*</span><span>4</span>)<span>*</span>(<span>5</span><span>*</span><span>6</span>)))<span>*</span>((<span>7</span><span>*</span><span>8</span>)<span>*</span><span>9</span>))</span>
<span><a href="#cb14-3"></a> &#9484;<span>1</span></span>
<span><a href="#cb14-4"></a> &#9484;&#9508;</span>
<span><a href="#cb14-5"></a> &#9474;&#9492;<span>2</span></span>
<span><a href="#cb14-6"></a>&#9484;&#9508;</span>
<span><a href="#cb14-7"></a>&#9474;&#9474; &#9484;<span>3</span></span>
<span><a href="#cb14-8"></a>&#9474;&#9474;&#9484;&#9508;</span>
<span><a href="#cb14-9"></a>&#9474;&#9474;&#9474;&#9492;<span>4</span></span>
<span><a href="#cb14-10"></a>&#9474;&#9492;&#9508;</span>
<span><a href="#cb14-11"></a>&#9474; &#9474;&#9484;<span>5</span></span>
<span><a href="#cb14-12"></a>&#9474; &#9492;&#9508;</span>
<span><a href="#cb14-13"></a>&#9474; &#9492;<span>6</span></span>
<span><a href="#cb14-14"></a>&#9508;</span>
<span><a href="#cb14-15"></a>&#9474; &#9484;<span>7</span></span>
<span><a href="#cb14-16"></a>&#9474;&#9484;&#9508;</span>
<span><a href="#cb14-17"></a>&#9474;&#9474;&#9492;<span>8</span></span>
<span><a href="#cb14-18"></a>&#9492;&#9508;</span>
<span><a href="#cb14-19"></a> &#9492;<span>9</span></span></code></pre>
</div>
<p>It does <em>not</em> build up the tree as balanced as it possibly could, though:</p>
<div>
<pre><code><span><a href="#cb15-1"></a><span>&gt;&gt;&gt;</span> (treeFold (<span>:*:</span>) <span>.</span> <span>fmap</span> <span>Leaf</span>) [<span>1</span>,<span>2</span>,<span>3</span>,<span>4</span>,<span>5</span>,<span>6</span>]</span>
<span><a href="#cb15-2"></a>((<span>1</span><span>*</span><span>2</span>)<span>*</span>((<span>3</span><span>*</span><span>4</span>)<span>*</span>(<span>5</span><span>*</span><span>6</span>)))</span>
<span><a href="#cb15-3"></a> &#9484;<span>1</span></span>
<span><a href="#cb15-4"></a>&#9484;&#9508;</span>
<span><a href="#cb15-5"></a>&#9474;&#9492;<span>2</span></span>
<span><a href="#cb15-6"></a>&#9508;</span>
<span><a href="#cb15-7"></a>&#9474; &#9484;<span>3</span></span>
<span><a href="#cb15-8"></a>&#9474;&#9484;&#9508;</span>
<span><a href="#cb15-9"></a>&#9474;&#9474;&#9492;<span>4</span></span>
<span><a href="#cb15-10"></a>&#9492;&#9508;</span>
<span><a href="#cb15-11"></a> &#9474;&#9484;<span>5</span></span>
<span><a href="#cb15-12"></a> &#9492;&#9508;</span>
<span><a href="#cb15-13"></a> &#9492;<span>6</span></span></code></pre>
</div>
<p>There&rsquo;s four elements in the right branch, and two in the left in the above example. Three in each would be optimal.</p>
<p>Wait&mdash;optimal in what sense, exactly? What do we mean when we say one tree is more balanced than another? Let&rsquo;s say the &ldquo;balance factor&rdquo; is the largest difference in size of two sibling trees:</p>
<div>
<pre><code><span><a href="#cb16-1"></a><span>balFac ::</span> <span>Tree</span> a <span>-&gt;</span> <span>Integer</span></span>
<span><a href="#cb16-2"></a>balFac <span>=</span> <span>fst</span> <span>.</span> go <span>where</span></span>
<span><a href="#cb16-3"></a><span> go ::</span> <span>Tree</span> a <span>-&gt;</span> (<span>Integer</span>, <span>Integer</span>)</span>
<span><a href="#cb16-4"></a> go (<span>Leaf</span> _) <span>=</span> (<span>0</span>, <span>1</span>)</span>
<span><a href="#cb16-5"></a> go (l <span>:*:</span> r) <span>=</span> (lb <span>`max`</span> rb <span>`max`</span> <span>abs</span> (rs <span>-</span> ls), rs <span>+</span> ls) <span>where</span></span>
<span><a href="#cb16-6"></a> (lb,ls) <span>=</span> go l</span>
<span><a href="#cb16-7"></a> (rb,rs) <span>=</span> go r</span></code></pre>
</div>
<p>And one tree is more balanced than another if it has a smaller balance factor.</p>
<p>There&rsquo;s effectively no limit on the balance factor for the typewriter method: when the input is one larger than a power of two, it&rsquo;ll stick the one extra in one branch and the rest in another (as with <code>[<span>1</span><span>..</span><span>9</span>]</code> in the example above).</p>
<p>For the slalom method, it looks like there&rsquo;s something more interesting going on, limit-wise. I haven&rsquo;t been able to verify this formally (yet), but from what I can tell, a tree of height will have at most a balance factor of the th <a href="https://oeis.org/A001045">Jacobsthal number</a>. That&rsquo;s (apparently) also the number of ways to tie a tie using turns.</p>
<p>That was just gathered from some quick experiments and <a href="https://oeis.org/">oeis.org</a>, but it seems to make sense intuitively. Jacobsthal numbers are defined like this:</p>
<div>
<pre><code><span><a href="#cb17-1"></a>j <span>0</span> <span>=</span> <span>0</span></span>
<span><a href="#cb17-2"></a>j <span>1</span> <span>=</span> <span>1</span></span>
<span><a href="#cb17-3"></a>j n <span>=</span> j (n<span>-</span><span>1</span>) <span>+</span> <span>2</span> <span>*</span> j (n<span>-</span><span>2</span>)</span></code></pre>
</div>
<p>So, at the top level, there&rsquo;s the imbalance caused by the second-last <code>pairFold</code>, plus the imbalance caused by the third-to-last. However, the third-to-last imbalance is twice what it was at that level, because it is now working with an already-paired-up list. Why isn&rsquo;t the second last imbalance also doubled? Because it&rsquo;s counteracted by the fact that we turned around: the imbalance is in an element that&rsquo;s a leftover element. At least that&rsquo;s what my intuition is at this point.</p>
<p>The minimum balance factor is, of course, one. Unfortunately, to achieve that, I lost some of the properties of the previous folds:</p>
<h2>Lengths</h2>
<p>Up until now, I have been avoiding taking the length of the incoming list. It would lose a lot of laziness, cause an extra traversal, and generally seems like an ugly solution. Nonetheless, it gives the most balanced results I could find so far:</p>
<div>
<pre><code><span><a href="#cb18-1"></a><span>treeFold ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> <span>NonEmpty</span> a <span>-&gt;</span> a</span>
<span><a href="#cb18-2"></a>treeFold f (x<span>:|</span>xs) <span>=</span> go (<span>length</span> (x<span>:</span>xs)) (x<span>:</span>xs) <span>where</span></span>
<span><a href="#cb18-3"></a> go <span>1</span> [y] <span>=</span> y</span>
<span><a href="#cb18-4"></a> go n ys <span>=</span> f (go m a) (go (n<span>-</span>m) b) <span>where</span></span>
<span><a href="#cb18-5"></a> (a,b) <span>=</span> <span>splitAt</span> m ys </span>
<span><a href="#cb18-6"></a> m <span>=</span> n <span>`div`</span> <span>2</span></span></code></pre>
</div>
<p><code><span>splitAt</span></code> is an inefficient operation, but if we let the left-hand call return its unused input from the list, we can avoid it:</p>
<div>
<pre><code><span><a href="#cb19-1"></a><span>treeFold ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> <span>NonEmpty</span> a <span>-&gt;</span> a</span>
<span><a href="#cb19-2"></a>treeFold f (x<span>:|</span>xs) <span>=</span> <span>fst</span> (go (<span>length</span> (x<span>:</span>xs)) (x<span>:</span>xs)) <span>where</span></span>
<span><a href="#cb19-3"></a> go <span>1</span> (y<span>:</span>ys) <span>=</span> (y,ys)</span>
<span><a href="#cb19-4"></a> go n ys <span>=</span> (f l r, rs) <span>where</span></span>
<span><a href="#cb19-5"></a> (l,ls) <span>=</span> go m ys</span>
<span><a href="#cb19-6"></a> (r,rs) <span>=</span> go (n<span>-</span>m) ls</span>
<span><a href="#cb19-7"></a> m <span>=</span> n <span>`div`</span> <span>2</span></span></code></pre>
</div>
<p>Finally, you may have spotted the state monad in this last version. We can make the similarity explicit:</p>
<div>
<pre><code><span><a href="#cb20-1"></a><span>treeFold ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> <span>NonEmpty</span> a <span>-&gt;</span> a</span>
<span><a href="#cb20-2"></a>treeFold f (x<span>:|</span>xs) <span>=</span> evalState (go (<span>length</span> (x<span>:</span>xs))) (x<span>:</span>xs) <span>where</span></span>
<span><a href="#cb20-3"></a> go <span>1</span> <span>=</span> state (\(y<span>:</span>ys) <span>-&gt;</span> (y,ys))</span>
<span><a href="#cb20-4"></a> go n <span>=</span> <span>do</span></span>
<span><a href="#cb20-5"></a> <span>let</span> m <span>=</span> n <span>`div`</span> <span>2</span></span>
<span><a href="#cb20-6"></a> l <span>&lt;-</span> go m</span>
<span><a href="#cb20-7"></a> r <span>&lt;-</span> go (n<span>-</span>m)</span>
<span><a href="#cb20-8"></a> <span>return</span> (f l r)</span></code></pre>
</div>
<p>And there you have it: three different ways to fold in a more balanced way. Perhaps surprisingly, the first is the fastest in my tests. I&rsquo;d love to hear if there&rsquo;s a more balanced version (which is lazy, ideally) that is just as efficient as the first implementation.</p>
<p>I have found two other uses for these folds other than simply constructing more balanced binary trees. The first is summation of floating-point numbers. If you sum floating-point numbers in the usual way with <code>foldl'</code> (or, indeed, with an accumulator in an imperative language), you will see an error growth of , where is the number of floats you&rsquo;re summing.</p>
<p>A well-known solution to this problem is the <a href="https://en.wikipedia.org/wiki/Kahan_summation_algorithm">Kahan summation algorithm</a>. It carries with it a running compensation for accumulating errors, giving it error growth. There are two downsides to the algorithm: it takes four times the number of numerical operations to perform, and isn&rsquo;t parallel.</p>
<p>For that reason, it&rsquo;s often not used in practice: instead, floats are summed <em>pairwise</em>, in a manner often referred to as <a href="https://en.wikipedia.org/wiki/Pairwise_summation">cascade summation</a>. This is what&rsquo;s used in <a href="https://github.com/numpy/numpy/pull/3685">NumPy</a>. The error growth isn&rsquo;t quite as good&mdash;&mdash;but it takes the exact same number of operations as normal summation. On top of that:</p>
<p>Dividing a fold into roughly-equal chunks is exactly the kind of problem encountered when trying to parallelize certain algorithms. Adapting the folds above so that their work is performed in parallel is surprisingly easy:</p>
<div>
<pre><code><span><a href="#cb21-1"></a><span>splitPar ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> (<span>Int</span> <span>-&gt;</span> a) <span>-&gt;</span> (<span>Int</span> <span>-&gt;</span> a) <span>-&gt;</span> <span>Int</span> <span>-&gt;</span> a</span>
<span><a href="#cb21-2"></a>splitPar f <span>=</span> go</span>
<span><a href="#cb21-3"></a> <span>where</span></span>
<span><a href="#cb21-4"></a> go l r <span>0</span> <span>=</span> f (l <span>0</span>) (r <span>0</span>)</span>
<span><a href="#cb21-5"></a> go l r n <span>=</span> lt <span>`par`</span> (rt <span>`pseq`</span> f lt rt)</span>
<span><a href="#cb21-6"></a> <span>where</span></span>
<span><a href="#cb21-7"></a> lt <span>=</span> l (n<span>-</span>m)</span>
<span><a href="#cb21-8"></a> rt <span>=</span> r m</span>
<span><a href="#cb21-9"></a> m <span>=</span> n <span>`div`</span> <span>2</span></span>
<span><a href="#cb21-10"></a></span>
<span><a href="#cb21-11"></a><span>treeFoldParallel ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> <span>NonEmpty</span> a <span>-&gt;</span> a</span>
<span><a href="#cb21-12"></a>treeFoldParallel f xs <span>=</span></span>
<span><a href="#cb21-13"></a> treeFold <span>const</span> (splitPar f) xs numCapabilities</span></code></pre>
</div>
<p>The above will split the fold into <code>numCapabilities</code> chunks, and perform each one in parallel. <code>numCapabilities</code> is a constant defined in <a href="https://hackage.haskell.org/package/base-4.10.0.0/docs/GHC-Conc.html">GHC.Conc</a>: it&rsquo;s the number of threads which can be run simultaneously at any one time. Alternatively, you could the function include a parameter for how many chunks to split the computation into. You could also have the fold adapt as it went, choosing whether or not to spark based on how many sparks exist at any given time:</p>
<div>
<pre><code><span><a href="#cb22-1"></a><span>parseq ::</span> a <span>-&gt;</span> b <span>-&gt;</span> b</span>
<span><a href="#cb22-2"></a>parseq a b <span>=</span></span>
<span><a href="#cb22-3"></a> runST</span>
<span><a href="#cb22-4"></a> (bool (par a b) (<span>seq</span> a b) <span>&lt;$&gt;</span></span>
<span><a href="#cb22-5"></a> unsafeIOToST (liftA2 (<span>&gt;</span>) numSparks getNumCapabilities))</span>
<span><a href="#cb22-6"></a></span>
<span><a href="#cb22-7"></a><span>treeFoldAdaptive ::</span> (a <span>-&gt;</span> a <span>-&gt;</span> a) <span>-&gt;</span> a <span>-&gt;</span> [a] <span>-&gt;</span> a</span>
<span><a href="#cb22-8"></a>treeFoldAdaptive f <span>=</span></span>
<span><a href="#cb22-9"></a> Lazy.treeFold</span>
<span><a href="#cb22-10"></a> (\l r <span>-&gt;</span></span>
<span><a href="#cb22-11"></a> r <span>`parseq`</span> (l <span>`parseq`</span> f l r))</span></code></pre>
</div>
<p>Adapted from <a href="https://www.reddit.com/r/haskell/comments/73umrw/another_parallelism_primitive_parseq/dnurduu/?utm_content=permalink&amp;utm_medium=front&amp;utm_source=reddit&amp;utm_name=haskell">this</a> comment by Edward Kmett. This is actually the fastest version of all the folds.</p>
<p>All of this is provided in a <a href="https://hackage.haskell.org/package/treefold">library</a> I&rsquo;ve put up on Hackage.</p>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via doisinkidney.com https://ift.tt/3m2f4eN<br>
<br>
October 20, 2020 at 04:22PM
